core/jvm/src/main/scala/zio/internal/ZScheduler.scala

```diff
@@ -417,35 +418,34 @@ private final class ZScheduler(autoBlocking: Boolean) extends Executor {
         }
       }
 
-      def markAsBlocking(workerIdx: Int, ignoreIfAutoBlocking: Boolean): Unit =
-        if (blocking || (ignoreIfAutoBlocking && autoBlocking)) ()
+      // NOTE: Synchronized block in case the supervisor attempts to mark the worker as blocking at the same time
+      // as an external call
+      def markAsBlocking(): Unit = synchronized {
+        if (blocking) ()
         else {
           blocking = true
-          val workerId =
-            if (workerIdx >= 0) workerIdx
-            else {
-              val i = workers.indexOf(self)
-              if (i >= 0) i else throw new Error("Catastrophic error: worker not found in workers array")
+          val idx = workers.indexOf(self)
+          if (idx >= 0) {
+            val runnables = self.localQueue.pollUpTo(256)
+            globalQueue.offerAll(runnables)
+            val worker = cache.poll()
```

Comment:
this may be a waste, shouldn't we consider the current workload before adding workers?
wouldn't the scheduler self-balance itself if the workload justifies this? otherwise it's quite possible it can comfortably operate with less workers atm and later allocate or reactivate a cached worker

Reply:
Look, I agree with you that we can do this in a better / more efficient way. The problem I'm seeing though is that this might be adding too much complexity for non-standard operations. E.g., we should be avoiding unsafely running effects within other effects like the plague. But users might rarely need to do it, so we need to account for it

Reply:
If we were to optimize it in the way that you're suggesting, we would need to rewrite the scheduling logic. This will end up adding some overhead to thread scheduling, and potentially introducing bugs. If this was a common use-case, I'd like to optimize more for it as well, but given that so far no one complained about the previous performance (which was relying on auto-blocking which has a 100ms delay till it marks the code as blocking), I think we can afford the overhead here

Reply:
when I approached you about this I was mainly thinking about documenting this and adding a sample that wraps the nested effect with `ZIO.blocking(...)`, that's what I did (luckily enough only in testing code).
I did also mention the possibility of auto-detecting this, and I tend to agree this is not something deserving a 'die-hard' approach 😎 

I do think that in this case when we're transferring work items from a worker specific queue to the global queue it's quite easy to detect if we significantly change the load on the scheduler and make a local decision if the scheduler requires more workers or not (is it limited to one?). I think this is a local decision and not something that changes the entire management of the scheduler.